Commands API

FelipeDefensor opened the pull request:
Introduces the commands API, a much simpler way to handle user interactions. Commands make all the following obsolete:

- All timeline and timeline component `RequestHandlers`
- `TimelineComponentSelector`
-  Several `TimelineSelectors`
- `TiliaActions`
- `user_action` fixture
- `TimelineUi.on_timeline_request`, `TimelineUI.on_timeline_component_request` (substituted by `TimelineUIs.on_timeline_command`)
- `TimelineUIs.pre_process_timeline_request()`
- `tilia\ui\timelines\collection\requests\args.py`
- Several `Post` and `Get` members
- Some `TimelineSelector` members

The core functionality is given by `command.py`, along with fuller documentation.

Usage of commands:
```
    # Register a new command
    commands.register(
        'example.command',
        callback_function,
        text='Menu Text',  # Optional: for display in Qt interface
        shortcut='Ctrl+E',  # Optional: keyboard shortcut
        icon='example_icon'   # Optional: icon name in IMG_DIR (without extension)
    )

    # Execute a command
    commands.execute('example.command', arg1, arg2, kwarg1=value1)
```

Examples of commands:
- File commands:        'file.open', 'file.save', 'file.export.img'
- Timeline commands:    'timeline.delete', 'timeline.component.copy'
- View commands:        'view.zoom.in', 'view.zoom.out'

Note that we can pass arguments to commands directly, so we don't need (horrible) things like `TimelineSelector.FROM_MANAGE_TIMELINES_TO_PERMUTE`.

Commands also set the stage for plugins, as they provide a simpler interface to modify application state programatically.

This PR also contains small changes that were covenient preliminaries for or complements to the implementation:
	- Bug fixes (3be8cee41664cc6157d9bb775ca0bf9f495fa942, 24576bddb619ab9e8c9ca3559988aa0e85621846, 6964fdd66faa38910addf783eac7710ffc37db30)
	- Additional testing (467cb5ed048ae5443218caa366ec066164c6980b, f07fd4da9304bac1695986d4ceeabce24e5edf3b)
	- Convenient refactors (b8d15ef49cfcb0cb904de54c0c9db368a0d62dfd,  914a0bd3a1cc925d644abe6540fca32f730cfb84, 974e71508f46ed562fc854746f0c9b68c6f1bc6a, 769e30ba35b73b763eb91cb08de12c2c00a0d462, f13fa08eb1255014fdb8033398c0b7213fc78287)
	- Documentation (45dd1dadace511835b260b1f694094e5e95d83b4)

FelipeDefensor committed:
refactor: hardcoded values

FelipeDefensor committed:
chore: auto-format

FelipeDefensor committed:
refactor: remove unused code

FelipeDefensor committed:
refactor: use Timeline.flags instead of individual class attributes

FelipeDefensor committed:
refactor: start using {Timeline,TimelineUI}.subclasses instead of timeline kinds

We should leverage classes as first-class citizens to use them instead of relying on an additional TimelineKind abstraction (TimelineKind). This is a first step towards that.

FelipeDefensor committed:
fix: do not usable mutable default argument

FelipeDefensor committed:
fix: crash when creating score annotation

FelipeDefensor committed:
fix: restore state with identical components but different ids

FelipeDefensor committed:
fix: crash when loading svg viewer without beat timeline

This is a hacky, temporary solution. This state can be achieved by:
- Creating a score timeline and then deleting the beat timeline used to create it
- Restoring the app state (e.g. after an exception during a timeline command), if the svg viewer happens to be loaded before the beat timeline is restored

FelipeDefensor committed:
test: set timeline visibility

FelipeDefensor committed:
fix: color validator

"" is a valid value for the color property, it means "use the default color", just as None.

FelipeDefensor committed:
fix: detection of empty timelines

Also add tests for clear commands

FelipeDefensor committed:
refactor: take "time" parameters to on_add method of point-like TimelineUI subclasses

FelipeDefensor committed:
test: record state after creating timelines with fixtures

FelipeDefensor committed:
refactor: rename Post.APP_RECORD_STATE to Post.APP_STATE_RECORD

FelipeDefensor committed:
refactor: commands APIS [MAJOR]

This commits introduces the commands API, a much simpler way to handle user interactions. Commands make all the following obsolete:

- All timeline and timeline component `RequestHandlers`
- `TimelineComponentSelector`
-  Several `TimelineSelectors`
- `TiliaActions`
- `user_action` fixture
- `TimelineUi.on_timeline_request`, `TimelineUI.on_timeline_component_request` (substituted by `TimelineUIs.on_timeline_command`)
- `TimelineUIs.pre_process_timeline_request()`
- `tilia\ui\timelines\collection\requests\args.py`
- Several `Post` and `Get` members
- Some `TimelineSelector` members

The core functionality is given by `command.py`, along with fuller documentation.

Usage of commands:
```
    # Register a new command
    commands.register(
        'example.command',
        callback_function,
        text='Menu Text',  # Optional: for display in Qt interface
        shortcut='Ctrl+E',  # Optional: keyboard shortcut
        icon='example_icon'   # Optional: icon name in IMG_DIR (without extension)
    )

    # Execute a command
    commands.execute('example.command', arg1, arg2, kwarg1=value1)
```

Examples of commands:
- File commands:        'file.open', 'file.save', 'file.export.img'
- Timeline commands:    'timeline.delete', 'timeline.component.copy'
- View commands:        'view.zoom.in', 'view.zoom.out'

Note that we can pass arguments to commands directly, so we don't need (horrible) things like `TimelineSelector.FROM_MANAGE_TIMELINES_TO_PERMUTE`.

Commands also set the stage for plugins, as they provide a simpler interface to modify application state programatically.

FelipeDefensor committed:
test: fix errors on test_dirs.py

FelipeDefensor committed:
fix: duplicated error message

Also improve message wording.

FelipeDefensor committed:
fix: error message wording

FelipeDefensor committed:
docs: add debug tip

FelipeDefensor committed:
fix: undo/redo when importing

FelipeDefensor committed:
refactor: remove dead code

FelipeDefensor committed:
fix: error message on failed command with a partial for a callback

FelipeDefensor committed:
fix: toggle timeline visibility

FelipeDefensor committed:
dev: clearer "Unregistered command" error output

FelipeDefensor committed:
fix: preserve timeline ids when opening files and undoing/redoing

FelipeDefensor committed:
fix: do not crash on race condition for creation of dirs.SITE_DATA_DIR

FelipeDefensor committed:
test: use deepdiff to print mismatching app states

FelipeDefensor committed:
fix: crash when using timeline context menus

FelipeDefensor committed:
fix: crash when deleting beat timeline after score timeline has been created

We should find a more robust solution for this. For instance, we should warn the user that the score won't scroll automatically if the beat timeline is deleted; and we shouldn't rely on dictionaries if we care about order.

FelipeDefensor committed:
refactor: remove CLI command

FelipeDefensor committed:
refactor: more consistent folder structure for CLI tests

FelipeDefensor committed:
feat: output warning in CLI

FelipeDefensor committed:
feat: raise warning when getting timelines by a duplicated name

FelipeDefensor committed:
refactor: move `permute_ordinal` to backend

FelipeDefensor committed:
refactor: add cli.io.error function

FelipeDefensor committed:
fix: "//" in YouTube URL being replaced with "/" when saving

FelipeDefensor committed:
docs: improve CLI help strings

azfoo commented on tests/ui/timelines/test_timeline_ui.py:
```diff
@@ -276,3 +277,15 @@ def test_set_empty_string(self, tls, tluis, user_actions):
+    with undoable():
+        commands.execute("timeline.set_is_visible", marker_tlui, True)
+
+    assert marker_tlui.view.isVisible() is True
```
assert not isVisible() / assert isVisible() ?

FelipeDefensor committed:
docs: mention CLI in feature list

FelipeDefensor committed:
docs: add tutorial for CLI

FelipeDefensor committed:
feat: add move CLI command

FelipeDefensor committed:
refactor: use refactored function

FelipeDefensor committed:
refactor: use simpler API for tests

FelipeDefensor committed:
refactor: use patch_yes_or_no instead of Serve

azfoo commented:
A locally run Linux pytest fails on the same test. We might have to look into why.

FelipeDefensor committed:
fix: media path = `.` when no media is loaded

FelipeDefensor committed:
fix: crash when timeline folder doesn't have the right structure

This crash was particularly annoying when we are working on a new timeline in a separate branch, because the timeline folder wont't get deleted by git when switching branches.

FelipeDefensor committed:
test: fix failing tests on Linux and macOS

Takeaway: we should use `QTest` to simulate interaction with the ui instead of calling widget methods attributes ourselves.

FelipeDefensor committed:
test: fix failing tests on Linux and macOS

Takeaway: we should use `QTest` to simulate interaction with the ui instead of calling widget methods attributes ourselves.

azfoo committed:
refactor: remove unused posts

azfoo committed:
refactor: actually call the post

azfoo commented on tilia/timelines/base/timeline.py:
```diff
@@ -33,11 +36,27 @@
+        TimelineFlag.HAS_COMPONENTS,
+        TimelineFlag.COMPONENTS_COPYABLE,
+        TimelineFlag.COMPONENTS_IMPORTABLE,
+    ]
```
Could this be made into something more generic so not every timeline rewrites this?

FelipeDefensor replied:
I didn't understand the comment. Could you be more specific. I am not entirely satisfied with the solution I came up with. A clear improvement would be to use [`enum.Flag`](https://docs.python.org/3/library/enum.html#enum.Flag) instead of a list.

azfoo replied:
Every timeline has rewritten this variable. I would prefer it if it were left as an empty list, or as a list of values that is shared by a number of them (probably copyable and importable).

PS. Is `TimelineFlag.HAS_COMPONENTS` used for anything?

FelipeDefensor replied:
Got it. Makes sense. Fixed in 0b2fca43.

PS. Fixed in 3251835b.

azfoo committed:
refactor: remove duplicate

azfoo committed:
refactor: remove unused gets

azfoo committed:
refactor: remove unused tl selectors

azfoo commented on tilia/timelines/base/timeline.py:
```diff
@@ -110,6 +132,31 @@ def components(self):
     def default_height(self):
         return None
 
+    @classmethod
+    def subclasses(cls):
+        cls.ensure_subclasses_are_available()
+        return cls.__subclasses__()
```
Can this be refactored so that `ensure_subclasses_are_available()` is only instantiated once and not every time the method is called?

FelipeDefensor replied:
This was only called when a new timeline was instantiated, so the performance impact is negligible. Even then, it was an easy fix in 09d3d764.

azfoo commented on tilia/timelines/base/validators.py:
```diff
@@ -17,7 +17,7 @@ def validate_bool(value):
 
 def validate_color(value):
-    if value is None:
+    if value is None or value == "":
```
is `if not value` not sufficient?

FelipeDefensor replied:
Not sure if I did this intentionally, but I guess it defends against other falsey values, which we don't want to see there. Are you against leaving it there?

azfoo replied:
more so that the acceptable types of QColor inits are known. a try/catch might work better because we could potentially crash in the next part.

FelipeDefensor replied:
How about what I did in 72bcd074?

azfoo commented on tilia/timelines/collection/collection.py:
```diff
@@ -299,12 +299,20 @@ def crop_timeline_components(self, new_length: float) -> None:
         tl = self.get_beat_timeline_for_measure_calculation()
+        if not tl:
+            return None
         beats = tl.components
```
Will this if clause not already be caught by `get_beat_timeline_for_measure_calculation()`?

FelipeDefensor replied:
I guess this is supposed to be merely defensive, but I may have got a crash on the `beats = tl.components` line the other day. Should have documented it.

FelipeDefensor replied:
I am leaving this just in case, ok?

azfoo commented on tilia/timelines/timeline_kinds.py:
```diff
@@ -40,23 +22,19 @@ def get_timeline_kind_from_string(string):
+    return kind.name.replace("_TIMELINE", "").lower()
+
+
+def get_timeline_frontend_name(kind: TimelineKind) -> str:
+    name = get_timeline_name(kind)
+    if kind == TimelineKind.PDF_TIMELINE:
+        name = name.upper()
+
+    return name
+
+
```
Where is this method used?

azfoo commented:
Just to clarify, why refactor some requests into commands but not the others?

azfoo committed:
refactor: correct method name

`on_side_arrow_press` and `on_horizontal_arrow_press` are duplicates, with the latter being defined in the parent tlui.
Redefined in audiowave because elements don't have "time" attr.

azfoo committed:
fix: restore missing inspector value

azfoo commented on tilia/ui/timelines/base/context_menus.py:
After adding a new timeline, additional move up/down options do not appear
e.g. create a new file, add marker tl1, then marker tl2. (so from top to bottom: slider, tl1, tl2). right-clicking on tl1 should have options for both up and down, but only up is shown. however, once moved up, the menu options are correct.

p.s. should slider tl also have a context menu with up/down options?

azfoo commented on tilia/ui/timelines/base/timeline.py:
```diff
@@ -84,9 +108,25 @@ def __bool__(self):
     def __lt__(self, other):
         return self.get_data("ordinal") < other.get_data("ordinal")
 
+    @classmethod
+    def subclasses(cls):
+        cls.ensure_subclasses_are_available()
+        return cls.__subclasses__()
+
+    @classmethod
+    def ensure_subclasses_are_available(cls):
+        timelines_dir = Path(os.path.dirname(__file__)).parent
+        packages = [
+            "tilia.ui.timelines." + d.name
+            for d in timelines_dir.iterdir()
+            if d.is_dir() and d.name not in ["base", "__pycache__"]
+        ]
+        for pkg in packages:
+            importlib.import_module(pkg)
+
```
as above in timelines/base/timeline

azfoo committed:
refactor: actually delete the file

azfoo commented on tilia/ui/timelines/collection/collection.py:
```diff
@@ -116,6 +126,274 @@ def _setup_widgets(self, main_window: QMainWindow):
+        except Exception:
+            tilia.errors.display(
+                tilia.errors.COMMAND_FAILED,
+                f"Timeline command callback '{callback.__name__}'",
```
`callback` might not have `__name__` if it turns out to not be a method

FelipeDefensor replied:
Fixed in 0793fd58

azfoo commented on tilia/ui/timelines/collection/collection.py:
```diff
@@ -116,6 +126,274 @@ def _setup_widgets(self, main_window: QMainWindow):
+            post(
+                Post.APP_STATE_RECORD,
+                f"timeline command: {method_name or callback.__name__}",
+            )
```
if `validate_command_return_value` only checks that the returned result is a list of bools / a bool and not if the command is actually successful, what happens after a failed command (result has a false) when the user tries to undo?

FelipeDefensor replied:
This was completely wrong. Thanks. Fixed by a9654aaf.

azfoo commented on tilia/ui/timelines/collection/collection.py:
```diff
@@ -116,6 +126,274 @@ def _setup_widgets(self, main_window: QMainWindow):
+        **kwargs,
+    ):
+        """
+        To be used as a wrapper for a timeline command.
```
but it isn't a wrapper?

FelipeDefensor replied:
Fixed in eff94b91

FelipeDefensor committed:
feat: preserve label and comments from merged units

FelipeDefensor committed:
fix: comments icon visibility

azfoo commented on tilia/ui/commands.py:
```diff
@@ -0,0 +1,146 @@
+def _execute_dev(command_name: str, *args, **kwargs):
+    if command_name not in _name_to_callback:
+        for key in _name_to_callback:
+            print(key)
```
can we not use print but instead attach it to the error message?

FelipeDefensor replied:
This has been improved in another branch. I'll find the commit and move it here.

FelipeDefensor replied:
Fixed by 23c2eda5

azfoo commented on tests/parsers/csv/test_markers_from_csv.py:
```diff
@@ -118,39 +113,41 @@ def test_markers_by_time_from_csv_fails_if_no_time_column(marker_tlui):
 
-    assert status == "success"
-    assert_in_errors("nonsense", errors)
+    assert success is True
```
assert success?

FelipeDefensor replied:
Fixed in df61c274

azfoo committed:
refactor: remove unused methods

azfoo committed:
refactor: typing

azfoo committed:
docs: typo

FelipeDefensor committed:
test: add helper functions for saving and opening tilia files

FelipeDefensor committed:
fix: duplicated IDs when using previously saved files

azfoo commented on tilia/app.py:
```diff
@@ -275,11 +275,30 @@ def on_record_state(self, action, no_repeat=False, repeat_identifier=""):
+        # Increments counter until a unique id is found.
+        # Could easily be optimised.
+        while str(next_id) in existing_ids:
+            next_id = next(self._id_counter)
```
How does this perform when creating a new element after loading a very large file?
I was wondering about setting the start id (`itertools.count(start)`) after getting the largest id from after the file loads. That would save having to loop through everything each time you get an id.
id will keep growing but that might be helpful when comparing across versions.
Also of potential interest is a way to validate ids when loading a file.

FelipeDefensor replied:
The performance is acceptable. In a file with thousands of hierarchies it just adds a ~1.5 second overhead to the ~12 seconds it was taking to open. Starting from the largest id like you suggested is more elegant and performant, though. It is also easy to implement. Will do that.

FelipeDefensor replied:
Validating IDs could be good, yes. I am considering implementing some major changes to IDs for v0.7, so let's leave this for then, ok?

FelipeDefensor replied:
Fixed in 26ebdf7e

azfoo committed:
docs: add help hints

azfoo committed:
docs: correct command

azfoo committed:
refactor: reorder

items show up alphabetically in help

azfoo committed:
fix: restore deleted get

initially deleted because it was unused

FelipeDefensor committed:
fix: dynamically set audio outputs

Before this changes to default audio outputs in the OS would be ignored. If the default output was removed, for instance, there was no way to regain sound playback, even if the output was added again.

FelipeDefensor committed:
fix: remove instructions for running CLI from binary

FelipeDefensor committed:
refactor: boolean comparison

azfoo committed:
docs: point to cli help

FelipeDefensor committed:
dev: improve unregistered command error message

FelipeDefensor committed:
refactor: boolean comparison

FelipeDefensor committed:
dev: improve docstring

FelipeDefensor committed:
perf: id generation

FelipeDefensor committed:
fix: crash when timeline command callback was not a function

FelipeDefensor committed:
fix: invalid return values for timeline commands not being detected

FelipeDefensor committed:
perf: import timeline subclasses only once

FelipeDefensor committed:
refactor: remove unused TimelineFlag

FelipeDefensor committed:
refactor: always set TimelineFlags explicitly

FelipeDefensor committed:
fix: color value validator
